geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/LayerMetadataStore.java

```diff
@@ -0,0 +1,358 @@
+/**
+ * This program is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU Lesser General Public License as published by the Free Software Foundation, either version 3
+ * of the License, or (at your option) any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * <p>You should have received a copy of the GNU Lesser General Public License along with this
+ * program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author Jose Macchi / Geosolutions 2009
+ */
+package org.geowebcache.storage.blobstore.file;
+
+import static org.geowebcache.storage.blobstore.file.FilePathUtils.filteredLayerName;
+
+import com.google.common.hash.Hashing;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStreamWriter;
+import java.io.UncheckedIOException;
+import java.io.UnsupportedEncodingException;
+import java.io.Writer;
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.geowebcache.util.FileUtils;
+import org.geowebcache.util.SuppressFBWarnings;
+
+public class LayerMetadataStore {
+
+    private static Log log =
+            LogFactory.getLog(org.geowebcache.storage.blobstore.file.LayerMetadataStore.class);
+
+    public static final String PROPERTY_METADATA_MAX_RW_ATTEMPTS =
+            "gwc.fileblobstore.maxRWAttempts";
+
+    static final int METADATA_MAX_RW_ATTEMPTS =
+            Integer.parseInt(System.getProperty(PROPERTY_METADATA_MAX_RW_ATTEMPTS, "10"));
+
+    static final String METADATA_GZIP_EXTENSION = ".gz";
+
+    private final String path;
+
+    private File tmp;
+
+    /** number of locks, make it configurable maybe */
+    private static final int lockShardSize = 32;
+
+    /** handling of local-process concurrent access to layer metadata files */
+    private ReadWriteLock[] locks =
```

Comment:
This helps in single instance, especially under load, but won't do a thing when using multiple GWC instances, on different nodes, concurrently accessing the same storage.